src/town_cmd.cpp

```diff
@@ -517,6 +517,50 @@ static void MakeTownHouseBigger(TileIndex tile)
 	if (flags & BUILDING_HAS_4_TILES) MakeSingleHouseBigger(TILE_ADDXY(tile, 1, 1));
 }
 
+/**
+ * Generate cargo for a house, scaled by the current economy scale.
+ * @param t The current town.
+ * @param ct Type of cargo to generate, usually CT_PASSENGERS or CT_MAIL.
+ * @param amount The number of cargo units.
+ * @param stations Available stations for this house.
+ * @param affected_by_recession Is this cargo halved during recessions?
+ */
+static void TownGenerateCargo(Town *t, CargoID ct, uint amount, StationFinder &stations, bool affected_by_recession)
+{
+	uint cargo_amount = amount;
+
+	/* All production is halved during a recession (except for NewGRF-supplied town cargo). */
+	if (affected_by_recession && EconomyIsInRecession()) {
+		cargo_amount = (cargo_amount + 1) >> 1;
+	}
+
+	/* If we're not producing any cargo, we can bail out now. */
+	if (cargo_amount == 0) return;
+
+	/* Maybe scale production by the chosen economy scale. */
+	if (_settings_game.economy.cargo_scale != 100) {
+		cargo_amount = ScaleByCargoScale(cargo_amount);
+	}
+
+	/* Actually generate cargo and update town statistics. */
+	const CargoSpec* cs = CargoSpec::Get(ct);
+	switch (cs->town_effect) {
+		case TE_PASSENGERS:
+			t->supplied[CT_PASSENGERS].new_max += cargo_amount;
+			t->supplied[CT_PASSENGERS].new_act += MoveGoodsToStation(CT_PASSENGERS, cargo_amount, ST_TOWN, t->index, stations.GetStations());
+			break;
+
+		case TE_MAIL:
+			t->supplied[CT_MAIL].new_max += cargo_amount;
+			t->supplied[CT_MAIL].new_act += MoveGoodsToStation(CT_MAIL, cargo_amount, ST_TOWN, t->index, stations.GetStations());
+			break;
+
+		default:
+			MoveGoodsToStation(ct, cargo_amount, ST_TOWN, t->index, stations.GetStations());
+			break;
+		}
+}
```

Comment:
I don't actually understand where this code is coming from; it feels off.

First of, in the old code it doesn't check town_effect. So that is new. And doesn't seem to be part of this PR.
Second, `CT_PASSENGERS` and `CT_MAIL` update `supplied`. But the old code also does it for the other cargoes.

Can you elaborate a bit what is happening here exactly? Tnx :)

Reply:
That's a separate change and a separate commit (or even PR) then. But I don't see anything with TownEffect here so maybe it's removed already.